cql3/expr/expression.hh

```diff
@@ -95,10 +98,17 @@ enum class comparison_order : char {
 
 /// Operator restriction: LHS op RHS.
 struct binary_operator {
-    std::variant<column_value, column_value_tuple, token> lhs;
+    std::unique_ptr<expression> lhs;
     oper_t op;
     ::shared_ptr<term> rhs;
-    comparison_order order = comparison_order::cql;
+    comparison_order order;
```

Comment:
I'm really curious (i.e., would have been nice to have a comment) what is this "comparison order thing" and why it's part of a binary operator.

Reply:
It's literally 10 lines above...

Reply:
I didn't see it in the github presentation of the patch (I thought it was something that Avi added).
Ok, now I understand it but I can't say I like it very much: It sounds to me like there are two different "<" operators - the cql one, and the clustering one (by the way, I don't understand what the second one means... Does it mean the tokens are compared for partition key? or something else). So you can have two operators - LT_CQL and LT_CLUSTERING. The general flag you have there suggests it applies to *every* operator. But does it apply to "="? Does it apply to "+"? I guess the answer is no.

Reply:
> I don't understand what the second one means

See the patch that introduced SCYLLA_CLUSTERING_BOUND, it explains the reasons.

> The general flag you have there suggests it applies to every operator

Every _ordering_ operator.

Reply:
Here's a discussion that led to the introduction of SCYLLA_CLUSTERING_BOUND:
https://github.com/scylladb/scylla-tools-java/pull/224#discussion_r573760181

Reply:
> > I don't understand what the second one means
> 
> See the patch that introduced SCYLLA_CLUSTERING_BOUND, it explains the reasons.
> 
> > The general flag you have there suggests it applies to every operator
> 
> Every _ordering_ operator.
> 
> > Does it apply to "+"? I guess the answer is no.
> 
> Correct.  That's because there is no `+`.

CQL has a +, so that's not the reason.

Reply:
> CQL has a +

`binary_operator` doesn't.

Reply:
`binary_operator` surely will gain the ability to add/concatenate, or we won't be able to use expressions in update statements.

Reply:
@dekimir if this class is intended only for **comparison** operators, it should be called `comparison_operator`. If it is called `binary_operator` it suggests it will also support `+` et al. in the (not so far-away) future. And if it will support many different operators, adding the comparison order to all of them is really weird.
I still don't understand the "comparison order" thing needs to be part of the expression - and not knowledge in the **evaluator** of the expression (who knows if it's comparing a clustering column or not), or, alternatively, part of the operator itself (you'll have two versions of each of the operators <, <=, >, >= - but not for any other operator).

Reply:
One day soon we'll separate out comparison operators into their own class with the `order` member, and this confusion will be avoided.  Or we will add different operators for different orders.  Until then, I feel it's acceptable to have this in `binary_operator`, even if it clearly doesn't apply to some operators.

> I still don't understand the "comparison order" thing needs to be part of the expression

It defines the meaning of the expression (see [here](https://sourcegraph.com/github.com/scylladb/scylla/-/blob/test/cql/query_bounds_test.result#L23-55)).  Any reasonable use of the expression must know the order.